docs/content/guides/developer/app-examples/reviews-rating.mdx

```diff
@@ -0,0 +1,205 @@
+
+### Reviewers
+
+Reviewers are consumers of services that use the review system. Reviewers provide feedback in the form of comments that detail specific aspects of the service as well as a star rating to inform others. The reviews are rated, with the most effective reviews getting the highest rating. Service owners award the 10 highest rated reviews for their service. How often the rewards are distributed is up to the service owner's discretion; for example, it can be distributed once a day or once a month.
+
+### Review readers
+
+Review readers access reviews to make informed decisions on selecting services. Readers rate reviews by casting up or down votes. The review readers' ratings are factored into the algorithm that rates the reviews, with the authors of the highest-rated reviews getting awarded. Although it is not implemented as part of this guide, a proportion of the review reward may be awarded to review readers who has participated in rating reviews by casting their votes.
+
+### Moderators
+
+Moderators monitor content of the reviews and can delete any reviews that contain inappropriate content. Moderators also keep an eye on the number of reviews registered for a service to ensure that they do not grow indefinitely.
+
+The incentive mechanism is not implemented for this guide, but service owners can all pay into a pool that goes to moderators, on a rolling basis. People can stake moderators to influence what proportion of the reward each moderator gets, up to a limit (imagine how validators are staked on-chain), and moderator decisions are decided by quorum of stake weight. This is a way to ensure that moderators have incentives to do their job well.
+
+## How reviews are rated
+
+The reviews are rated (scored) on-chain using following criteria
+- Intrinsic score (IS)
+  - Length of content
+- Extrinsic score (ES)
+  - Number of votes received
+- Verification multiplier (VM)
+  - Reviews with PoE get a multiplier to improve rating
+
+```
+Total Score = (IS + ES) * VM
+```
+
+## Smart contracts
+
+There are several modules that create the backend logic for the example.
+
+### dashboard.move
+
+The `dashboard.move` module defines the `Dashboard` struct that is used to group services.
+
+```move
+struct Dashboard has key, store {
+  id: UID,
+  service_type: String
+}
+```
+
+The services are grouped by attributes, which can be cuisine type, geographical location, operating hours, google maps ID, and so on. To keep it basic, the example stores only `service_type` (for example, fast food, Chinese, Italian).
+
+### service.move
+
+This module defines the `Service` struct, which service owners manage.
+
+```move
+struct Service has key, store {
+  id: UID,
+  reward_pool: Balance<SUI>,
+  reward: u64,
+  top_reviews: MultiMap<ID>,
+  reviews: Table<ID, ID>,
```

Comment:
The flow of a review is not clear to me, e.g.

- how is a review created and inserted into these structures,
- how is it voted upon,
- how does it transition from the `reviews` structure to the `top_reviews` structure and vice versa,
- how do we guarantee that reviews are visible and can be voted upon by anybody (or to phrase this in a different way, how do we prevent bad actors from hiding undesirable reviews).
- how do we guarantee that a review that gets sufficient votes ends up in `top_reviews`?

Reply:
- A `review` is a shared object, so anyone can cast a vote on a `review` and recompute `total_score`. 
  - I've added a link to "shared object" Object Ownership documentation and also added more descriptions for casting votes.
- There is a function called `update_top_reviews` that updates `top_reviews` 
  - I've added `update_top_reviews` that inserts a new review into `top_reviews` after `total_score` is updated.

Reply:
IIUC, this means that when you create a review, you create a shared object, and then you put its ID in the `reviews` Table, then to update a review, the client would need to look it up in the `reviews` table (out of interest, what ID is used for the keys in the `reviews` table?), and then issue a separate transaction with the ID of the shared object containing the actual review. If that update means that it should now be in the top reviews, it needs to then remember to call `update_top_reviews` to perform that update on the `Service`.

This is not ideal, because it is non-atomic and allows mistakes, adversarial behaviour (imagine a particular restaurant monitors the sentiment of reviews as they are receiving votes and "forgets" to update the top reviews if the votes are for negative sentiment reviews), or timing issues on the client side to present an inconsistent state on-chain: You could have review objects that don't appear in the `top_reviews` structure that have more reviews that do.

The fix is to represent the `Service` as follows:

```move
struct Service has key, store {
  id: UID,
  reward_pool: Balance<SUI>,
  reward: u64,
  top_reviews: vector<ID>,
  reviews: ObjectTable<ID, Review>,
  overall_rate: u64,
  name: String,
}
```

Reviews are stored in an `ObjectTable`, meaning that they are stored as children of the shared object, but they are still accessible off-chain by their ID (that's the difference between `Table` and `ObjectTable`) and we re-use their ID as the key of the `reviews` table.

Because `top_reviews` now only needs to store ~10 reviews, and modifications to this structure should always happen at the end (the bottom review gets replaced), it can be stored in a simple `vector` (no more need for a custom data structure). This vector is stored in order of reviews, and holds the IDs of the top reviews (whose content and vote count can be looked up from the `reviews` table).

The function for casting a vote will now also check whether the review should be added to `top_reviews`. This guarantees atomicity, and no foul play.

Reply:
I've removed the custom data structure and used `ObjectTable` to store `reviews`. I also extended the section on casting votes which would always trigger calling `update_top_reviews` to order `top_reviews`.